src/VisualStudio/Core/Def/Telemetry/AggregatingTelemetryLog.cs

```
@@ -0,0 +1,125 @@
+        private readonly object _lock;
+        private readonly HistogramConfiguration? _histogramConfiguration;
+        private readonly string _eventName;
+
+        private const string MeterVersion = "0.39";
+
+        public AggregatingTelemetryLog(TelemetrySession session, FunctionId functionId, double[]? bucketBoundaries = null)
+        {
+            var meterName = TelemetryLogger.GetPropertyName(functionId, "meter");
+            var meterProvider = new VSTelemetryMeterProvider();
+
+            _session = session;
+            _meter = meterProvider.CreateMeter(meterName, version: MeterVersion);
+            _histograms = new();
+            _lock = new();
+            _eventName = TelemetryLogger.GetEventName(functionId);
+
+            if (bucketBoundaries != null)
+            {
+                _histogramConfiguration = new HistogramConfiguration(bucketBoundaries);
+            }
+        }
+
+        /// <summary>
+        /// Adds aggregated information for the metric and value passed in via logMessage. KeyValueLogMessage.Properties
+        /// is searched for appropriate entries to determine the metric name and value.
+        /// </summary>
+        /// <param name="logMessage"></param>
+        public void Log(LogMessage logMessage)
+        {
+            if (!IsEnabled)
+                return;
+
+            if (logMessage is not KeyValueLogMessage kvLogMessage)
+                return;
+
+            string? metricName = null;
+            int? value = null;
+            foreach (var kvp in kvLogMessage.Properties)
+            {
+                switch (kvp.Value)
+                {
+                    case string s:
+                        metricName = s;
+                        break;
+                    case int i:
+                        value = i;
+                        break;
+                }
+            }
+
+            if (metricName is null || value is null)
+                return;
```

Is this an error case because we are silently ignoring this log call? Should we throw `ExceptionUtilities.Unreachable` here instead or this can happen in error cases and we want to handle it gracefully?

---

commit 3